js/misc/missionGameService.js

```diff
@@ -0,0 +1,267 @@
+// -*- mode: js; js-indent-level: 4; indent-tabs-mode: nil -*-
+
+const Lang = imports.lang;
+const GLib = imports.gi.GLib;
+const Gio = imports.gi.Gio;
+const Showmehow = imports.gi.Showmehow;
+const Signals = imports.signals;
+
+
+/**
+ * callWhenComplete
+ *
+ * Takes an object with keys mapping to bool values, a callback
+ * and a key name. Flips the key name to true and checks if
+ * they are all true, and if so, calls the callback.
+ *
+ * This is used by MissionChatboxTextService to ensure that
+ * all asynchronous tasks are completed before calling
+ * the completion callback. The callback is called with
+ * the key that triggered the callback.
+ */
+function callWhenComplete(completion, key, callback) {
```

Comment:
There is code with a similar concept at https://github.com/endlessm/eos-desktop/blob/master/js/gdm/batch.js -- would be nice if we could reuse that